Add maisi controlnet training

guopengf opened the pull request:
Fixes # .

### Description
Add training script for MAISI ControlNet.

### Checks
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Avoid including large-size files in the PR.
- [ ] Clean up long text outputs from code cells in the notebook.
- [ ] For security purposes, please check the contents and remove any sensitive info such as user names and private key.
- [ ] Ensure (1) hyperlinks and markdown anchors are working (2) use relative paths for tutorial repo files (3) put figure and graphs in the `./figure` folder
- [ ] Notebook runs automatically `./runner.sh -t <path to .ipynb file>`

mingxin-zheng commented:
Hi @guopengf , can we have some doc-string added in the functions? Thanks!

mingxin-zheng commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,405 @@
+
+
+def setup_ddp(rank, world_size):
+    print(f"Running DDP diffusion example on rank {rank}/world_size {world_size}.")
```
Since we are using the `logging` modules, shall we change the `print` to `logging.info`?

Also, the RankFilter in `monai` may be helpful to print message in selected rank (e.g. rank=0) in dist training case.

guopengf replied:
thanks, refactored code to use logging and monai RankFilter.

guopengf commented:
> Hi @guopengf , can we have some doc-string added in the functions? Thanks!

Hi @mingxin-zheng, thanks for the comments. I added docstring for util functions.

mingxin-zheng commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,427 @@
+
+    # Step 2: define diffusion model and controlnet
+    # define diffusion Model
+    unet = define_instance(args, "difusion_unet_def").to(device)
```
The variable name `difusion_unet_def` seems like it could be a typo. It should likely be `diffusion_unet_def`.

guopengf replied:
Thanks, I corrected it.

mingxin-zheng commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,427 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+from argparse import Namespace
+import json
+import logging
+from pathlib import Path
+import time
+from datetime import timedelta
+from typing import Any
+import os
+import sys
+import copy
+import torch
+import torch.distributed as dist
+import torch.nn.functional as F
+import logging
+from monai.utils import RankFilter
+from monai.data import DataLoader, CacheDataset, partition_dataset
+from monai.networks.utils import copy_model_state
+from torch.cuda.amp import GradScaler, autocast
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.utils.tensorboard import SummaryWriter
+from monai.transforms import (
+    Compose,
+    EnsureTyped,
+    Lambdad,
+    LoadImaged,
+    Orientationd,
+)
+from monai.bundle import ConfigParser
+from utils import binarize_labels
```
```suggestion
import argparse
import copy
import json
import logging
import os
import sys
import time
from argparse import Namespace
from datetime import timedelta
from pathlib import Path
from typing import Any

import torch
import torch.distributed as dist
import torch.nn.functional as F
from monai.bundle import ConfigParser
from monai.data import CacheDataset, DataLoader, partition_dataset
from monai.networks.utils import copy_model_state
from monai.transforms import Compose, EnsureTyped, Lambdad, LoadImaged, Orientationd
from monai.utils import RankFilter
from torch.cuda.amp import GradScaler, autocast
from torch.nn.parallel import DistributedDataParallel as DDP
from torch.utils.tensorboard import SummaryWriter
from utils import binarize_labels
```

Please run isort <filename> to sort the imports. Thanks!

mingxin-zheng commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,427 @@
+    # define diffusion Model
+    unet = define_instance(args, "difusion_unet_def").to(device)
+    # load trained diffusion model
+    diffusion_model_ckpt = torch.load(args.trained_diffusion_path, map_location=device)
```
Consider adding error handling for loading checkpoints (`torch.load`) in case of issues such as file not found, incorrect format,

guopengf replied:
Thanks for the suggestion. I added the error handling.

mingxin-zheng commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,427 @@
+        data_base_dir=args.data_base_dir,
+        rank=rank,
+        world_size=world_size,
+        batch_size=args.controlnet_train["batch_size"],
```
Please also make sure the args are parsed correctly. I didn't find it directly in the argument list.

guopengf replied:
Those arguments are stored in configs/config_maisi_controlnet_train.json. https://github.com/Project-MONAI/tutorials/blob/40526e05fc5e2ec6d370c51de7f36f5426704edb/generative/maisi/configs/config_maisi_controlnet_train.json#L70

mingxin-zheng commented:
@KumoLiu I don't have any further comments except the ones above.

KumoLiu commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,427 @@
+    # Step 0: configuration
+    logger = logging.getLogger("maisi.controlnet.training")
+    # whether to use distributed data parallel
+    ddp_bool = args.gpus > 1
+    if ddp_bool:
```
```suggestion
    use_ddp = args.gpus > 1
    if use_ddp:
```

KumoLiu commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,427 @@
+        LoadImaged(keys=["image", "label"], image_only=True, ensure_channel_first=True),
+        Orientationd(keys=["label"], axcodes="RAS"),
+        EnsureTyped(keys=["label"], dtype=torch.uint8, track_meta=True),
+        Lambdad(keys="top_region_index", func=lambda x: torch.FloatTensor(x)),
+        Lambdad(keys="bottom_region_index", func=lambda x: torch.FloatTensor(x)),
+        Lambdad(keys="spacing", func=lambda x: torch.FloatTensor(x)),
```
I believe these three functions can be combined to one `Lambdad`. And also for Line153-Line155.

guopengf replied:
Thanks, replace them with one Lambdad.

guopengf commented:
Hi @KumoLiu and @mingxin-zheng, thanks for the valuable comments! I addressed them in the latest commit. In addition, I moved utility functions used in the controlnet training script to scripts/utils.py since I found Can's PR has been merged.

KumoLiu commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,274 @@
+
+    # Step 1: set data loader
+    train_loader, _ = prepare_maisi_controlnet_json_dataloader(
+        json_data_list=args.json_data_list,
```
I didn't find where define this json_data_list, did you run the training script locally?

guopengf replied:
Yeah, I tested it locally. `json_data_list` is defined in the environment config. https://github.com/Project-MONAI/tutorials/blob/3234841a5b44562359cdb3ecb2f00d7cc7564e42/generative/maisi/configs/environment_maisi_controlnet_train.json#L10

KumoLiu replied:
Yes, I mean that maybe we need some instruction here to tell users how to download the data?
Such as check if the json file and data dir is not exist, download them using `download_and_extract`.
https://github.com/Project-MONAI/MONAI/blob/3a0c2d5afcc91379819fd8428fa83b00fb18c9fd/monai/apps/utils.py#L304

What do you think?

guopengf replied:
Yeah, we do need instructions. I will add instructions about the controlnet training and data in another PR (https://github.com/Project-MONAI/tutorials/pull/1743).

KumoLiu commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,274 @@
+    parser.add_argument(
+        "-c",
+        "--config-file",
+        default="./config/config_maisi_controlnet_train.json",
```
```suggestion
        default="./configs/config_maisi_controlnet_train.json",
```

KumoLiu commented on generative/maisi/scripts/train_controlnet.py:
```diff
@@ -0,0 +1,274 @@
+    parser.add_argument(
+        "-e",
+        "--environment-file",
+        default="./config/environment_maisi_controlnet_train.json",
```
```suggestion
        default="./configs/environment_maisi_controlnet_train.json",
```
